odc/geo/_xr_interop.py

```diff
@@ -245,6 +246,96 @@ def assign_crs(
     return xx
 
 
+def mask(xx: XrT, poly: Geometry, all_touched: bool = True) -> XrT:
+    """
+    Apply a polygon geometry as a mask, setting all xr.Dataset
+    or xr.DataArray pixels outside the rasterized polygon to ``NaN``.
+
+    :param xx:
+       :py:class:`~xarray.Dataset` or :py:class:`~xarray.DataArray`.
+
+    :param poly:
+       Geometry shape used to mask ``xx``.
+
+    :param all_touched:
+        If ``True``, all pixels touched by ``poly`` will remain unmasked.
+        If ``False``, only pixels whose center is within the polygon or
+        that are selected by Bresenham's line algorithm will remain unmasked.
+
+    :return:
+        A :py:class:`~xarray.Dataset` or :py:class:`~xarray.DataArray`
+        masked by ``poly``.
+    """
+
+    # Reproject `poly` to match `xx`
+    poly = poly.to_crs(crs=xx.odc.crs)
+
+    # Rasterise `poly` into geobox of `xx`
+    rasterized = rasterize(
+        poly=poly,
+        how=xx.odc.geobox,
+        all_touched=all_touched,
+    )
+
+    # Mask data outside rasterized `poly`
+    xx_masked = xx.where(rasterized.data)
```

Comment:
So, `xx.where` in that form always changes output to `float{64|32}+nan` representation. Which is a valid and often desired output format, but not always. One might also expect to retain existing pixel data type with `nodata` on output.

We should certainly support both. But it's less clear to me which one should be a default

* Personally I would expect to get `int16, nodata=-9999` on output if that was given on input, but I can see how `float32, nan` and even `float64, nan` can be also desired.
* But I can also see that some would expect and prefer to see `float32, nan` output for `mask` operation. 

In the code as it stands we should at least ensure that `nodata` attribute is not present on `xx_masked` output array, I still don't have any intuition for `xarray` attributes, when they stay and when they go...

Reply:
Yeah, this occurred to me too - I definitely think both options should be supported. My take would be that "most" of our more general users simply want a way to quickly set pixels outside of a polygon to `NaN` - which is a super common workflow for things like pixel drills or zonal statistics. Getting back `int16, nodata=-9999` is definitely useful but perhaps for slightly more advanced users/workflows, so I would be in favour of the simpler `NaN` option being the default.

The approach taken in `dea_tools.datahandling.load_ard` seems like it could be a good template:
- `dtype="auto"`: When 'auto' is used, the data will be converted to `float32` if masking is used, otherwise data will be returned in the native data type of the data.
- `dtype="native"`: Data will be returned in the native data type of the data, including native nodata value
- `dtype='float{16|32|64}'": Custom dtype
https://github.com/GeoscienceAustralia/dea-notebooks/blob/develop/Tools/dea_tools/datahandling.py#L214-L221

Reply:
For the moment, I've added some code to strip nodata attributes:
https://github.com/opendatacube/odc-geo/blob/crop/odc/geo/_xr_interop.py#L281-L286